Issue 3619 - Allow control of recursion for includeAll via minDepth and maxDepth attributes

jasonlyle88 opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes expected existing functionality)
- [X] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Adds two new attributes to the changelog `includeAll` attribute: `minDepth` (inclusive, default `1`) and `maxDepth` (inclusive, default `Integer.MAX_VALUE`). These defaults allow `includeAll` to behave as it did before these provided changes.

For `minDepth`, `0` and `1` are treated the same way and denote the directory specified in the `includeAll`'s `path` attribute. Having a `minDepth` of 2 would skip any files specified in the `includeAll`'s `path` attribute and only start searching for files to include in immediate subdirectories of the `includeAll`'s `path` attribute.

For `maxDepth`, `1` is the minimum value. A value less than `minDepth` will result in no files being included. The `maxDepth` attribute is computed off of the `includeAll`'s `path` attribute, not based on the `minDepth` attribute.

Examples:

`minDepth`=`1`,`maxDepth`=`1`: Include all the files in the specified `path` attribute directory and do not look in any subdirectories (identical to recursive: false).

`minDepth`=`1`,`maxDepth`=`2`: Include all the files in the specified `path` attribute directory and look in immediate subdirectories and those subdirectories' immediate subdirectories.

... So on and so forth for maxDepth

`minDepth`=`2`,`maxDepth`=`null`: Do not include any files in the specified `path` attribute directory and look in all subdirectories for files.

## Things to be aware of

Not only did this require the change of the `includeAll` attribute, but also the `ResourceAccessor` interface.

As this change occurred on the base interface, all `ResourceAccessors` changed to support this model.

## Things to worry about

1. I've never worked with the Spring framework before. Because of this, I as unsure how/if I should handle the depth bounded search function for the `SpringResourceAccessor` class. For now, I created the method and just immediately throw an error saying it is not implemented. Not sure if this is acceptable or if this should be fully implemented.
2. I wasn't sure how/if I should implemented the depth bounded search function for the `MockResourceAccessor` class. For now, I created the method and just immediately throw an error saying it is not implemented. Not sure if this is acceptable or if this should be fully implemented.
3. I am unsure how to properly write tests for this when it is based on a filesystem configuration.

nvoxland commented on liquibase-core/src/main/java/liquibase/resource/ResourceAccessor.java:
```diff
@@ -112,6 +112,30 @@ default SortedSet<String> list(String relativeTo, String path, boolean recursive
+     * @return empty set if nothing was found
+     * @throws IOException if there is an error searching the system.
+     */
+    List<Resource> search(String path, Integer minDepth, Integer maxDepth) throws IOException;
```
Having a new search() version that is smarter allows us to add nice functionality like you have here.

However, just adding this new search function to the interface will break any extensions which simply implement this directly. I also see an unending set of other options people will want to add to the search logic over time which will lead to a large number of methods that ResourceAccessor implementations will have to think about and callers have to decide between.

I'm thinking it would be better to have ResourceAccessor.SearchOptions inner class with fields of minDepth, maxDepth, and recursive and change this new method to be `search(String path, SearchOptions searchOptions)` instead. That way future new like "file name pattern" or whatever can be included without having to make new method permutations.

Then, I'm wondering if we could default that new search method to call the existing `search(path, boolean)` call and do in-memory filtering of the results? That would allow ResourceAccessor implementors to generally just have the one method to implement and keep getting the additional functionality as it's added? But still leaving them to implement the new `search(String, SearchOptions)` if they feel they can do a better job?

jasonlyle88 replied:
I really like that idea, the more extensible the better! Especially in this situation where things are community driven and there is this extension capability. I do like the default implementation in the interface to allow it to move forward without breaking current implementations and allow future implementors to override only if they want.

I can definitely take a stab at that... however I am not a Java programmer. I picked this up as we are using Liquibase for our project and this is the first time I've touched Java in ~15 years (and I've only been using Liquibase for ~ 1 month). As such, please feel free to let me know where improvements can be made as my Java knowledge is very outdated and I'm still learning Liquibase!

jasonlyle88 commented:
@nvoxland I believe I have implemented the `SearchOptions` class how you were describing it. I also implemented the method in the `ResourceAccessor` interface as a default method so that any existing/future extensions don't have any issues after this is merged. 

Here are some updates to my original post for this merge request:

# Things to be aware of

I believe this default `search(String, SearchOptions)` method in the interface strategy takes care of the items I mentioned above in the "Things to be aware of" section.

# Things to worry about

1. `SpringResourceAccessor` still does not have this implemented. I don't think thats an issue... but wanted to make sure I called it out at least.
2. `MockResourceAccessor` has the method implemented, but doesn't take any of the SearchOption attributes into account. It previously did not look at the `recursive` value either, so I figure this is the intended functionality, but again just want to call it out to be thorough.
3. Not knowing how to implement tests based on filesystem hierarchy is still a concern.
4. (NEW) I opted not to create a `recursive` attribute of this class because that is really just a specific expression of `minDepth` and `maxDepth`. I did create `getRecursive` and `setRecursive` methods, but those just set `minDepth` and `MaxDepth` to appropriate values. Let me know if this is an issue, I just saw it as a better way of doing this since they really tell you the same information and could just cause problems trying to keep `recursive` and `minDepth`/`maxDepth` in sync.

nvoxland commented:
Good point that `recursive` is just a limited version of min/max depth. No need to have that too. 

The changes is looking generally good, so sending it along for the deeper review. I like that for our ResourceAccessor implementations you have fixed them up to implement the new SearchOptions search. But I wonder if it's possible to have the default ResourceAccessor SearchOptions method you made not just call the old recursive search and ignore the min/max settings, but still respect them? Maybe take the list of Resources returned from the recursive search and loop through them and pull out resources too high or low in the tree? Like what you did in `residesInDepthBoundaries` but with the more limited API of Resource. You might only have splitting on "/" or "\" but that would still better respect those settings for ResourceAccessors that only implement the recursive search version.

jasonlyle88 commented:
Awesome, I'm hoping the review goes well. I was able to get a basic depth bounding into the default `search(String, SearchOptions)` implementation based off `Resource.getPath()` and counting the `/` characters since `Resource.getPath()` says it should always be relative and use `/` as the separator. Let me know if you see any problems with this implementation.

MalloD12 reviewed (approved):
**Approved**.

## Review and testing results:

Changes looks good to me. As part of the review we have performed the following actions:

- Fixed failing unit tests
- Remove attribute default values in XSD
- Remove attribute types in XSD
- Added validation errors when setting up minDepth and maxDepth
- Change types from integer to primitive int type
- Update minDepth initialisation to be set with 0, instead of 1.
- Update getRecursive and setRecursive methods.
- Add validation error to return exception if maxDepth is less than minDepth
- ResourceAccessor search logic updated to make proper use of `searchRecursive` flag.
- MockResourceAccessor updated to only make use of recursive search method. Also, search method has been updated to make a proper use of **recursive** boolean flag.
- Manual testing performed with below combinations:
> 0-0
> 0-1
> 1-0
> 1-1
> 0-2
> 1-2
> 0-3
> 3-123123

- Unit test added to test `includeAll` with new min/max boundaries.
- Integration tests updated to also make use of these new `includeAll` properties.

### Things to be aware of:
- None: this is only having an impact on includeAll tag, but this is still supporting previous includeAll definition (without needing to specify recursive boundaries).

### Things to worry about:
- None

jasonlyle88 commented:
@MalloD12 No comments here! I've been following along and everything has made sense. Thank you very much for your work on getting this across the finish line!

Only question I have is if this will make it into the next release. I see a merge freeze check in the git checks and wasn't sure.

filipelautert commented:
Hello @jasonlyle88  - we are going to release 4.19.1 this week but this PR won't be able to make it. As we try to release monthly the next release that it may make into is around beggining of April.

filipelautert commented on liquibase-core/src/main/java/liquibase/resource/ResourceAccessor.java:
```diff
@@ -257,4 +308,51 @@ public OutputStream openOutputStream(OpenOptions openOptions) throws IOException
+
+        public SearchOptions() {
+            minDepth = 0;
+            maxDepth = 1;
```
Should not the default value be  Integer.MAX_VALUE ? This way it will behave the same as before if new people just do `new SearchOptions() `

MalloD12 replied:
In my opinion, it wouldn't change that much when SearchOptions is initialized it will be followed up with min/maxDepth seters. It we do initialize it with `Integer.MAX_VALUE` it will still define whether it's a recursive search depending on the values set. To keep recursive compatibility as it was before this change, there is a `search()` which expect a `recursive` boolean flag instead of a SearchOptions (as well as an IncludeAll method which does not expect min/max values which will end up performing a recursive search `min=0` - `max=Integer.MAX_VALUE`)

filipelautert reviewed (commented):
thanks for the code! I just added one comment about the default values.

jasonlyle88 commented:
@nvoxland not sure if you have to approve the changes for this to move forward once the merge freeze is done, but figured I'd ask now and get the ball rolling in case.

I think everything you asked for is covered here though!

XDelphiGrl commented on liquibase-core/src/main/java/liquibase/changelog/DatabaseChangeLog.java:
```diff
@@ -995,4 +1005,27 @@ public enum OnUnknownFileFormat {
+        ResourceAccessor.SearchOptions searchOptions = new ResourceAccessor.SearchOptions();
+        try {
+            if (maxDepth < minDepth) {
+                throw new IllegalArgumentException("maxDepth argument must be greater than minDepth");
```
Nice error handling!

XDelphiGrl commented:
This ticket needs to be tested on Windows. Internal Liquibase QA will take this action item.

XDelphiGrl commented:
**Test Methodology: includeAll**
To validate various combinations of `includeAll` `minDepth` and `maxDepth` properties, create a set of nested directories. Each directory will have one XML changelog with one changeset, except for one nested directory which is empty. The top-level directory contains the "parent" YAML chaneglogs; each of these changelogs has one `includeAll` changetype with varying combinations of `minDepth` and `maxDepth`. Parent changelogs are named according to the settings for `minDepth` and `maxDepth`. 

This set of tests is executed on Windows 11.

---

**Directory Structure**

~/ISSUE3619

~/ISSUE6319/depth1
issue3619-1-changelog.xml

~/ISSUE6319/depth1/depth2
issue3619-2-changelog.xml

~/ISSUE3619/depth1/depth2/depth3
issue3619-3-changelog.xml

~/ISSUE3619/depth1/depth2/depth3/depth4-no-changelog

~/ISSUE3619/depth1/depth2/depth3/depth4-no-changelog/depth5
issue3619-5-changelog.xml

---
**Parent Changelogs**
* ISSUE3619/issue3619-none-changelog.yaml (no min or max configured)
* ISSUE3619/issue3619-min0max1-changelog.yaml
* ISSUE3619/issue3169-min1-changelog.yaml
* ISSUE3619/issue3619-min1max2-changelog.yaml
* ISSUE3619/issue3619-min1max2-depth3-changelog.yaml (relative nested path using /depth1/depth2/depth3)
* ISSUE3619/depth1/depth2/depth3/depth4-no-changelog/depth5/issue3619-min1-max2-relativedepth3-changelog.yaml (relative path using `..` to ISSUE3619/depth1/depth2/depth3)
* ISSUE3619/issue3619-min1max0-changelog.yaml

---
**Tests: includeAll**
Changelog for Cases 1, 2, 3, 4, 5 and 7 are executed from directory `ISSUE3619`.
Changelogs for Case 6 is executed from directory `ISSUE3619\depth1\depth2\depth3\depth4-no-changelog\depth5`
* Relative path is `/../../../depth3` 

| Case # | Parent  | Min | Max | Deployed Directories | Expected Count Changesets Deployed | P/F |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| 1 | issue3619-none | NOT SET  |  NOT SET | 1,2,3,4,5 | 4 | PASS |
| 2 | issue3619-min0max1 | 0  |  1 | 1 | 1 |  PASS |
| 3 |issue3619-min1 | 1  |  1 | 1,2,3,4,5 | 5 | PASS |
| 4 |issue3619-min1max2 | 1  | 2 | 1,2 | 2 | PASS |
| 5 |issue3619-min1max2-depth3 | 1 | 2  | 3 | 1 | PASS |
| 6 | issue3619-min1max2-relativedepth | 1 | 2 | 3 |  1 | PASS |
| 7 | issue3619-min1max0 | N/A  |  0 | ERROR | 0 | PASS |

---
**Tests: includeAll wrapped in modifyChangesets**

This test uses a parent changelogs with an `includeAll` wrapped in `modifyChangesets`.  Formatted SQL changelogs, each with one changeset are located in `ISSUE3619/depth1`, `ISSUE3619/depth1/depth2` and `ISSUE3619/depth1/depth2/depth3`. This test validates the following conditions:
* JSON `includeAll` supports `minDepth` and `maxDepth`
* XML `includeAll` supports `minDepth` and `maxDepth`
* An `includeAll` wrapped with `modifyChangesets` correctly uses the PSQL native exexutor (PRO feature).

**Example of includeAll Wrapped in modifyChangesets (JSON)**
```
   "modifyChangeSets": 
       {
           "runWith": "psql",
           "includeAll": {"path": "depth1", "relativeToChangelogFile": "true", "minDepth": "0", "maxDepth": "3"}
       }
```

Changelogs for Case 8 and Case 9 are executed from directory `ISSUE3619`.

| Case # | Parent  | Min | Max | Deployed Directories | Expected Count Changesets Deployed | P/F |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| 8 | issue3619-modifychangeset.json | 0 |  1 | 1,2,3 | 3 | PASS |
| 9 | issue3619-modifychangeset.xml | 0 |  1 | 1,2,3 | 3 | PASS |

XDelphiGrl commented on liquibase-core/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:
```diff
@@ -419,7 +443,7 @@ create view sql_view as select * from sql_table;'''
         ])
         def changeLogFile = new DatabaseChangeLog("com/example/root.xml")
-        changeLogFile.includeAll("com/example/missing", false, null, true, changeLogFile.getStandardChangeLogComparator(), resourceAccessor, new ContextExpression(), new Labels(), false)
+        changeLogFile.includeAll("com/example/missing", false, null, true, changeLogFile.getStandardChangeLogComparator(), resourceAccessor, new ContextExpression(), new LabelExpression(), false)
```
Shouldn't `LabelExpression` be `Labels`? As I understand it, the changelog cannot have a label expression (or filter). In looking at  [the includeAll case statement near this line in DatabaseChangelog.java](https://github.com/liquibase/liquibase/blob/63705dc90d775723d48859e33bdcbbb67c237572/liquibase-core/src/main/java/liquibase/changelog/DatabaseChangeLog.java#L519), it looks like Labels is the correct object type.  However ... this is working so maybe I just need some general education on when to use `Labels` versus `LabelExpression`.

MalloD12 replied:
These tests are using the deprecated version of `includeAll` :

```
public void includeAll(String pathName,
                           boolean isRelativeToChangelogFile,
                           IncludeAllFilter resourceFilter,
                           boolean errorIfMissingOrEmpty,
                           Comparator<String> resourceComparator,
                           ResourceAccessor resourceAccessor,
                           ContextExpression includeContextFilter,
                           LabelExpression labelExpression,
                           boolean ignore)
```
Which we still use for backward compatibility. These tests were failing because passing labels (instead of label expression) would also require including these new parameters (minDepth/maxDepth), so at that moment I thought might be simpler to keep it that way, but I can update these tests and make use of the a non-deprecated includeAll method.

XDelphiGrl commented on liquibase-core/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:
```diff
@@ -590,4 +614,32 @@ http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbch
+        rootChangeLog.getChangeSets().size() == expectedIncludeAllChangesetsToDeploy
+
+        where:
+        minDepth | maxDepth | expectedIncludeAllChangesetsToDeploy
```
This is a very nice set of test cases!

XDelphiGrl reviewed (approved):
This new feature allows you to control the depth of recursion applied by `includeAll`. To do so, set the `minDepth` and `maxDepth` properties on an `includeAll` changeset. Not setting either `minDepth` or `maxDepth` is the equivalent of pre-existing behavior and all directories are recursed to find changelogs to execute.
* This feature does not break backwards compatibility.
* This feature DOES have an XSD change and can only be released with a major version.
* This feature works with `includeAll` changesets wrapped in a `modifyChangesets` changetype; this is used for the PRO-only native executors, like SQLPlus, SQLCmd and PSQL.
* The Firebird test harness failures are due to an environmental issue with Titan and are not related to this PR. 
* The functional test failures are related to an issue fixed on the functional test master branch; again, these are not related to this PR and can be disregarded.
* Manual testing executed on Windows and Mac to validate there are no pathing issues related to this change.

APPROVED

jasonlyle88 commented:
Thanks everyone for your work here and helping me get this included! Very much appreciated!
